scripts/gen_vimdoc.py

```diff
@@ -38,37 +39,91 @@
   - Parameters are omitted for the `void` and `Error *` types, or if the
     parameter is marked as [out].
   - Each function documentation is separated by a single line.
+
+TODO:
+- Handle defs like 'vim.g' and others getting generated
+
+- Handle local functions being exported, even though they should not be
+    - I think I fixed this, but there is a new problem
+
+    ```lua
+    local function do_x()
+    end
+
+    return {
+      do_stuff = do_x
+    }
+    ```
 """
+
+import argparse
+import collections
 import os
 import re
-import sys
 import shutil
-import textwrap
 import subprocess
-import collections
-import msgpack
-
+import sys
+import textwrap
+from typing import Any
+from typing import Dict
 from xml.dom import minidom
 
+try:
+    import msgpack
+except ImportError:
+    print("package 'msgpack' is required to run the script", file=sys.stderr)
+    sys.exit(1)
+
 if sys.version_info[0] < 3 or sys.version_info[1] < 5:
-    print("requires Python 3.5+")
+    print("requires Python 3.5+", file=sys.stderr)
+    sys.exit(1)
+
+script_path = os.path.abspath(__file__)
+base_dir = os.path.dirname(os.path.dirname(script_path))
+out_dir = os.path.join(base_dir, "tmp-{target}-doc")
+filter_cmd = "{} {}".format(
+    sys.executable, os.path.join(base_dir, "scripts", "vim2dox_filter.py")
+)
+lua2dox_filter = os.path.join(base_dir, "scripts", "lua2dox_filter")
+
+if not shutil.which("doxygen"):
+    print("Missing Requirement: doxygen", file=sys.stderr)
     sys.exit(1)
 
-DEBUG = ('DEBUG' in os.environ)
-TARGET = os.environ.get('TARGET', None)
-INCLUDE_C_DECL = ('INCLUDE_C_DECL' in os.environ)
-INCLUDE_DEPRECATED = ('INCLUDE_DEPRECATED' in os.environ)
+if not any(
+    shutil.which(x)
+    for x in {
+        "./.deps/usr/bin/luajit",
+        "./.deps/usr/bin/lua",
+        "lua",
+        "luajit",
+        "texlua",
+    }
+):
+    print("Missing Requirement: Need lua or luajit", file=sys.stderr)
+    sys.exit(1)
+
+if subprocess.run([lua2dox_filter, "--test"]).returncode:
+    print("Failed to execute: ", f"$ {lua2dox_filter} --test", file=sys.stderr)
+    sys.exit(1)
+
+# TODO: Why are we just checking if they are in the environ or not,
+#       instead of checking their values?
```

Comment:
no reason except that this script was not supposed to become mission critical :P